qadence/mitigations/zne.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+
+from typing import cast
+
+import numpy as np
+import torch
+from pulser_simulation import SimConfig
+from torch import Tensor
+
+from qadence import BackendName
+from qadence.backend import ConvertedCircuit, ConvertedObservable
+from qadence.backends.api import backend_factory
+from qadence.backends.pulser.backend import Backend
+from qadence.blocks import block_to_tensor
+from qadence.measurements import Measurements
+from qadence.mitigations import Mitigations
+from qadence.types import NoiseModels
+from qadence.utils import Endianness
+
+
+def zne(noise_probas: list, zne_dataset: list) -> float:
+    # Rearrange the dataset by selecting each element in the batches.
+    # TODO: Correct for arbitrary batches.
+    rearranged_dataset = [s[0][0] for s in zne_dataset]
+    # Polynomial fit.
+    p = np.poly1d(np.polyfit(noise_probas, rearranged_dataset, 4))
+    return float(p(0.0))  # Return the zero-noise fitted value.
+
+
+def analog_zne(
+    backend_name: BackendName,
+    circuit: ConvertedCircuit,
+    observable: list[ConvertedObservable] | ConvertedObservable,
+    param_values: dict[str, Tensor] = {},
+    state: Tensor | None = None,
+    protocol: Measurements | None = None,
+    mitigation: Mitigations | None = None,
+    endianness: Endianness = Endianness.BIG,
+) -> Tensor:
+    assert mitigation
+    noise_model = mitigation.options.get("noise_model")
+    if noise_model is None:
+        KeyError(f"A noise model should be choosen from {NoiseModels.list()}. Got {noise_model}.")
+    assert noise_model
+    noise_probas = mitigation.options.get("noise_probas")
+    if noise_probas is None:
+        KeyError(f"A range of noise probabilies should be passed. Got {noise_probas}.")
+    noise_probas = cast(list, noise_probas)
+    backend = backend_factory(backend=BackendName.PULSER, diff_mode=None)
+    backend = cast(Backend, backend)  # Cast the Pulser backend.
+    backend_config = backend.config
+    # Construct the ZNE dataset.
+    zne_dataset = []
+    for noise_proba in noise_probas:
+        # Setting the backend config to account for the noise.
+        if noise_model == NoiseModels.DEPOLARIZING:
+            backend_config.sim_config = SimConfig(noise=noise_model, depolarizing_prob=noise_proba)
+        elif noise_model == NoiseModels.DEPHASING:
+            backend_config.sim_config = SimConfig(noise=noise_model, dephasing_prob=noise_proba)
+        # Get density matrices in the noisy case.
+        density_matrices = backend.run_noisy(
+            circuit, param_values=param_values, state=state, endianness=endianness
+        )
```

Comment:
I suggested to remove this `run_noisy`, now I see how you use it. To make it cleaner, I would define a `run_dm` function in the backend which returns a density matrix instead of the wavefunction. Only the Pulser backend would implement it, thus making this function fail if the Pulser backend is not used + it makes it also a bit more flexible for future updates (e.g. at some point also PyQ will implement a `run_dm` method).

Reply:
Instead of adding yet another method to the backends, would it make sense to add an argument to query for the output of run ? Say, `Output.WF` or `Output.DENSITY` ? That was also a query from yesterday's meeting. @madagra ?

Reply:
Or a `Result` type would be even cleaner I think.

Reply:
In #187